packages/sanity/src/core/store/_legacy/ResourceCacheProvider.tsx

```diff
@@ -16,7 +16,7 @@ export interface ResourceCacheProviderProps {
 
 /** @internal */
 export function ResourceCacheProvider({children}: ResourceCacheProviderProps) {
-  const resourceCache = useMemo((): ResourceCache => {
+  const [resourceCache] = useState((): ResourceCache => {
```

Comment:
A great rule of thumb for determining wether a `useMemo` is safe and side-effects free is wether you can remove it and what happens to your app:
- a) it still works, but slower
- b) it breaks, or is so much slower it's unusable

In this case it's b), as the assumption here is that `useMemo` with zero dependencies means `resourceCache` will run exactly once, and have a stable instance for hte lifecycle of components calling it. This is false. In fact, it can choose to throw away the memoized value in certain suspense scenarios: https://react.dev/reference/react/useMemo#caveats

A better solution is to use `useState`, as it does have a guarantee that it's never throwing away the value, and it has a lazy init function so that our function that creates the map is still only run once.

This is also reflected in how much simpler the React Compiler optimization is [for `useState`](https://playground.react.dev/#N4Igzg9grgTgxgUxALhASwLYAcIwC4AEwBeAnlggQEoICGceAchACYIA0BUYCAwhADs8CAB55O3BAGU8tYQQC+BAGYwIGAgHIYdBpoA6AzDnxFqCSLES96ACz6DhYxSrUbNYWkbIB6APpgaAIA5gA2CHiCYAYChsa4hMRwOnIIALJQoXhoANIIpADqdADWabRYnGQUBBlZuflFtKXlLqrqWgB0Psl0wrXZeYUlZVgxhj4AVBMEAAJBwjACtKEEEz6GoiaE8wgwyvSUNJbwfHaUwIYEBMERADwAKgQAvFwCxQIQAO4CAHwAFBAsNkosgzEsMBYsAdQWA8DAgsEANwENgUARsARwNAWUEAgBGACsEAwCAAfAgCTKhACUAG0ALqKamgx7kqDohDKIIIFiXAg8PAAoFoEFg2gQsBQxAwuEI5GohAczHYsC4iCE4mEcmU0I0hnIgBuyygCFB7PeXwETNBBogaF5AgUhnGU1mO0Wy1W6wEmwSBHd+0Q5mO1jOAAU1Ab7bsI4CwEQ+XBbGhQiwdAJQTR6ExWAhDE7YgJJtM5kJdksVmsNiItip2QwRVajtATjYkwhY1G2DA-klk6n04pMxYW6H253ozBY1gwNSE1aCHAooRaToQ6d24yXpIZKk-n9mcHRxv7M8fvOrlclwJYRTxZCDvGXj7PjUWrd+bKQpx+vUhk0RnPA9nQXK4fB8Ehk3jNB40kFgSAgAgdCwUIDkgygAAMdVCDCCCNUITWgq0vFIAgMIVJUsQsXDIEguQ+TAiC8HsUi4C8Ag8UoOD-StZjMN-QZGmaLBcM+ZMkxRCALAETRCHoRAgQpKkGMXZclN1GgUIOGBniIAsVJ0PBYCtC5QMvG48FxYhwQfRBOAojEqPjBQ5yec9TMvTzVJvQgHMVJyCjQZjoCYKldL85ULA6DByn3VFXJ+FSvJRBAsGeJ5nzCgB+dTQk01DEB00FUXYJLL2pMqrzUmzJQOEZdJqqUoosv5GoOCqzM8wzjLvCUmpGLKOhaiKAqC2wQsYKkOuShRSsLZKBSs3rbI4FK0UclVOHwk0mTPC9ku8282sQern3vWrEDAIaIla86mrnUlyR6VIBIaYZYumg7r1vEaVUC4KoFC3VwtSxUNqimKsDi1KEsqy9UXSzLgZy7D8u0ghitSuaDoIT7kuOqKBVuvqDk4AmRjxrzyfKDoid+ix-vGwHJt1LbjQQSmrlmlSCy5jq+W6xYCD+FTbmbKwTwcMsxA6Cduzw9mnmANdjzbewFB+YAkxTNNFQUW4fHF1szn4aW8FlyNJ0ShcOv0otXVLBYKy9ata2UetgStSQjbHewD2Hdc1fORM1LY9tdMkU2nEFH3Jaj0Q8E+tBlGFgBCMP7DnZi1FfF8CAAURgNQe00fhMngj5CC5dEyIzhBcOvaPNE+wWrTr-MQAUIA), [compared to the `useMemo`](https://playground.react.dev/#N4Igzg9grgTgxgUxALhASwLYAcIwC4AEwBeAnlggQEoICGceAchACYIA0BUYCAwhADs8CAB55O3BAFkEGCAQC+BAGYwIGAgHIYdBpoA6AzDnxFqCSLES96ACz6DhYxSrUbNYWkbIB6APpgaAIA5gA2CHiCYAYChsa4hMRwOrTCUlCheGgA0gikAOp0ANZStFicZBQE6Zk5eYW0JWUuqupaAHQ+yXRpGVm5BcWlWDGGPgBU4wQAAkHCMAK0oQTjPoaiJoRzCDDK9JQ0lvB8dpTAhgQEwREAPAAqBAC8XAJFAhAA7gIAfAAUEFgslFkGZFhgLFh9iCwHgYEFggBuAhsCgCNgCOBoCwg-4AIwAVggGAQAD4EAQZUIASgA2gBdRRUkEPMlQNEIZRBBAsC4EHh4f6AtDA0G0cFgSGIaGw+FIlEIdkYrFgHEQAlEwhkimhan0pEANyWUAQILZb0+AkZIP1EDQPIECkMY0mM22CyWKzWAg2CQIbr2iHMR2spwACmp9XaduGAWAiLy4LY0KEWDoBCCaPQmKwEIZHbEBBMprMhDtFstVusRJsVGyGMLLYdoMcbImEDHI2wYL8kkmU2nFBmLM2Q22O1GYDGsGAqfHLQQ4FFCDpgyc208uDwZHJfr8mUGR2v7E9vnPLpdFwIYeSxRD9nHnt6PtVmjc+TKQpwav16kMyqe9ydedLh8HwSCTOM0DjSQWBIeQdCwUJ9nAygAANtVCVCCENUJjUgy0vFIAhUPlRVMQsLDIHA1JeRAsC8HsIi4C8AhcUoGC-UtBi0O-OpBkaYYsI+JNE2RCALAETRCHoRBAXJSlaIXJd5J1GhEP2GAN2AfNFJ0PBYEtc5gPPa48BxYgwTvRBOFI9FyLjBRZ0eU8jPPNylKvQhbIVez8jQBjoCYSkN28pULHaDAyl3FEnO+RT3ORBAsCeR5H2CgB+FTQjUpDEE0kEUXYeLzypYqL2UyyJX2YYN0qyVwtM346v2UrjLcvSDJvcV6uGdL2ka0LfP82xAsYSlWoShQioLBL+XMrqrI4RLUTs5VOBw41GRPM8Eo869msQGrH1vKrEDAfqIiak76tnEkyW6VJpD6PiGiaLBALa89L2vQblT8gKoCCnUQqShVVvCyL3t+GKTzK88URStLgcyjCco0ggCqS6bdoICbdoO8L+Su7r9k4AnhjxhLybKdoid+ix-pGwGxp1dajQQSnzymxT80uKaCHpVreQ6hYCF+RSbibKwjwcUsxHaccu2w9nHmAFdD1bewFG+YBE2TVMFQUG4fClltTn4OW8AViMJzi+dWp0wsXRLeZy09Ksa2UOsgUtSRTdHew9yHVdNbOBNlOY9dnkkC2nAFf2Zdj0Q8DxtBlDFgBCSP7FnBi1GfJ8CAAURgNRu00fgMlg95CE5NFiOzhAsMvOPNDxkXLUbvMQAUIA).